src/OmniSharp.Host/FileWatching/ManualFileSystemWatcher.cs

```diff
@@ -1,24 +1,44 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.IO;
+using OmniSharp.Models.FilesChanged;
 
 namespace OmniSharp.FileWatching
 {
     public class ManualFileSystemWatcher : IFileSystemWatcher
     {
-        private readonly Dictionary<string, Action<string>> _callbacks = new Dictionary<string, Action<string>>();
+        private readonly Dictionary<string, Action<string, FileChangeType>> _callbacks = new Dictionary<string, Action<string, FileChangeType>>();
+        private readonly Dictionary<string, Action<string, FileChangeType>> _directoryCallBacks = new Dictionary<string, Action<string, FileChangeType>>();
 
-        public void TriggerChange(string path)
+        public void TriggerChange(string path, FileChangeType changeType)
         {
-            Action<string> callback;
-            if (_callbacks.TryGetValue(path, out callback))
+            if (_callbacks.TryGetValue(path, out var callback))
             {
-                callback(path);
+                callback(path, changeType);
+            }
+
+            var directoryPath = Path.GetDirectoryName(path);
+            if (_directoryCallBacks.TryGetValue(directoryPath, out var fileCallback))
+            {
+                fileCallback(path, changeType);
             }
         }
 
-        public void Watch(string path, Action<string> callback)
+        public void Watch(string path, Action<string, FileChangeType> callback)
         {
             _callbacks[path] = callback;
         }
+
+        public void WatchDirectory(string path, Action<string, FileChangeType> callback)
+        {
+            if (_directoryCallBacks.TryGetValue(path, out var existingCallback))
+            {
+                _directoryCallBacks[path] = (Action<string, FileChangeType>)Delegate.Combine(callback, existingCallback);
```

Comment:
I'm OK leaving this as is, but it's also possible to just write:

```C#
_directoryCallBacks[path] = callback + existingCallback;
```

Believe it or not, `+` handles combination of delegates in C#. :smile:

**The More You Know**

Reply:
So much nicer. Fixing.